flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/filemerging/FileMergingSnapshotManagerBase.java

```diff
@@ -0,0 +1,337 @@
+     * @return the generated file path for a physical file.
+     */
+    protected Path generatePhysicalFilePath(Path dirPath) {
+        // this must be called after initFileSystem() is called
+        // so the checkpoint directories must be not null if we reach here
+        final String fileName = UUID.randomUUID().toString();
+        return new Path(dirPath, fileName);
+    }
+
+    /**
+     * Delete a physical file by given file path. Use the io executor to do the deletion.
+     *
+     * @param filePath the given file path to delete.
+     */
+    protected final void deletePhysicalFile(Path filePath) {
+        ioExecutor.execute(
+                () -> {
+                    try {
+                        fs.delete(filePath, false);
+                        LOG.debug("Physical file deleted: {}.", filePath);
+                    } catch (IOException e) {
+                        LOG.warn("Fail to delete file: {}", filePath);
+                    }
+                });
+    }
+
+    // ------------------------------------------------------------------------
+    //  abstract methods
+    // ------------------------------------------------------------------------
+
+    /**
+     * Get a reused physical file or create one. This will be called in checkpoint output stream
+     * creation logic.
+     *
+     * <p>TODO (FLINK-32073): Implement a CheckpointStreamFactory for file-merging that uses this
+     * method to create or reuse physical files.
+     *
+     * <p>Basic logic of file reusing: whenever a physical file is needed, this method is called
+     * with necessary information provided for acquiring a file. The file will not be reused until
+     * it is written and returned to the reused pool by calling {@link
+     * #returnPhysicalFileForNextReuse}.
+     *
+     * @param subtaskKey the subtask key for the caller
+     * @param checkpointId the checkpoint id
+     * @param scope checkpoint scope
+     * @return the requested physical file.
+     * @throws IOException thrown if anything goes wrong with file system.
+     */
+    @Nonnull
+    protected abstract PhysicalFile getOrCreatePhysicalFileForCheckpoint(
+            SubtaskKey subtaskKey, long checkpointId, CheckpointedStateScope scope)
+            throws IOException;
+
+    /**
+     * Try to return an existing physical file to the manager for next reuse. If this physical file
+     * is no longer needed (for reusing), it will be closed.
+     *
+     * <p>Basic logic of file reusing, see {@link #getOrCreatePhysicalFileForCheckpoint}.
+     *
+     * @param subtaskKey the subtask key for the caller
+     * @param checkpointId in which checkpoint this physical file is requested.
+     * @param physicalFile the returning checkpoint
+     * @throws IOException thrown if anything goes wrong with file system.
+     * @see #getOrCreatePhysicalFileForCheckpoint(SubtaskKey, long, CheckpointedStateScope)
+     */
+    protected abstract void returnPhysicalFileForNextReuse(
```

Comment:
IIUC, returnPhysicalFileForNextReuse must be called after using a physical file ?
otherwise caller has to maintain the close logic of the physical file, right ?

Reply:
Yes, that's right.